Add docs for how web-features syncs with BCD (#2560)
web-platform-dx/web-features: docs/bcd-sync.md

# BCD synchronization

> The [web-platform-dx/web-features](https://github.com/web-platform-dx/web-features/) and the [mdn/browser-compat-data](https://github.com/mdn/browser-compat-data/) repositories depend on, and complement each other. This document describes the how the two projects get synchronized and how BCD and web-features maintainers should work with both repositories. If you are a consumer of web-features data, these internal steps are mostly opaque to you. If you are a web-features maintainer, read this document to decide how to map or tag BCD keys.

Reviewer: ```suggestion
The [web-platform-dx/web-features](https://github.com/web-platform-dx/web-features/) and the [mdn/browser-compat-data](https://github.com/mdn/browser-compat-data/) repositories depend on, and complement each other. This document describes how the two projects get synchronized and how BCD and web-features maintainers should work with both repositories. If you are a consumer of web-features data, these internal steps are mostly opaque to you. If you are a web-features maintainer, read this document to decide how to map or tag BCD keys.
```

## BCD keys in `compat_features`

The source YAML files, that define each feature in the web-features, might contain a `compat_features` field.

> When present, the `compat_features` field consists of a list if browser-compat-data (BCD) entry keys that make up this feature (e.g., `css.properties.background-color`).

Reviewer: ```suggestion
When present, the `compat_features` field consists of a list of browser-compat-data (BCD) entry keys that make up this feature (e.g., `css.properties.background-color`).
```

> If the `compat_features` field is not present in a feature's source YAML file, the resulting `.yml.dist` file (generated by the `dist` script) will contain a `compat_features` field that contains the BCD entry keys that were tagged with `web-features:<feature-id>` in the BCD repository, if any.

Reviewer 2: ```suggestion
If the `compat_features` field is not present in a feature's source YAML file, the resulting `.yml.dist` file (generated by the `dist` script) contains a `compat_features` field that contains the BCD entry keys that were tagged with `web-features:<feature-id>` in the BCD repository, if any.
```

> The `compat_features` field in a feature's source YAML file always takes precedence over tags defined in BCD.

Reviewer 2: ```suggestion
The `compat_features` field in a feature's source YAML file always takes precedence over tags defined in BCD.
Whatever the origin of the `compat_features` keys, the `.dist` file shows the final, resolved list of keys.
```

## Tagging keys in BCD

One way to map keys to web-features is that keys are tagged in BCD from inception. This is more likely for small, incremental changes to existing features.

For example, a typical scenario could look like this:

- a new key appears in BCD (e.g., via a [Collector](https://github.com/openwebdocs/mdn-bcd-collector) PR).
- The author of the BCD PR knew which feature the key belongs to (e.g., due to a key being renamed) and tags the key with the correct `web-features:*` tag.

> - A few days later, that data is released by BCD. That triggers the next `@mdn/browser-compat-data` Dependabot upgrade PR, where we regenerate web-features data based on the updated tag and remove the `compat_features` list from the YAML file (if the list and tagged keys match).

Reviewer 3: There are two repositories, it seems worth being always explicit on where PRs get created.
Also, the `compat-features` list is removed from which YAML file? The "source YAML file" or the "resulting YAML file"? Probably both as a build likely happens, but I suppose we're mostly interested in the source files here, and the sentence is not clear on whether these get updated.

```suggestion
- A few days later, that data is released by BCD. That triggers the next `@mdn/browser-compat-data` Dependabot upgrade PR in the web-features repository, where we regenerate web-features data based on the updated tag and remove the `compat_features` list from the source YAML file (if the list and tagged keys match).
```

> - a new key appears in BCD (e.g., via a [Collector](https://github.com/openwebdocs/mdn-bcd-collector) PR).
> - The author of the BCD PR knows which feature the key belongs to (e.g., due to a key being renamed) and tags the key with the correct `web-features:*` tag.
> - A few days later, that data is released by BCD. That triggers the next `@mdn/browser-compat-data` Dependabot upgrade PR in the web-features repository, where we regenerate web-features data based on the updated tag and remove the `compat_features` list from the source YAML file (if the list and tagged keys match).

Reviewer 2: Show that this is a linear sequence of events and indicate where we're going to automate:

```suggestion
1. A new key appears in BCD (e.g., via a [Collector](https://github.com/openwebdocs/mdn-bcd-collector) PR).
2. The author of the BCD PR knows which feature the key belongs to (e.g., due to a key being renamed) and tags the key with the correct `web-features:*` tag.
3. A few days later, that data is released by BCD. That triggers the next `@mdn/browser-compat-data` Dependabot upgrade PR in the web-features repository, where we regenerate web-features data based on the updated tag and automatically remove the redundant `compat_features` list from the source YAML file (if the list and tagged keys match).
```

 If necessary (e.g., a tag was assigned by BCD maintainers in error), a web-features maintainer can update a `compat_features` list in web-features and the data gets synchronized to BCD as described in the next section.

## Listing keys in web-features source YAML files

> The other way to map keys to web-features is to list them in the source YAML files. In this scenario, web-features receives unaccounted-for keys in the `features/draft` folder which will have to be looked into by web-features maintainers and will later be added to BCD as tags automatically.

Reviewer 2: ```suggestion
The other way to map keys to web-features is to list them in the source YAML files, which are eventually migrated to BCD as tags through an automated process. In this scenario, web-features receives unaccounted-for keys in the `features/draft` folder, which must be addressed by web-features maintainers.
```

For example, a typical scenario would look like this:

- a new key appears in BCD (e.g., via a [Collector](https://github.com/openwebdocs/mdn-bcd-collector PR).
- The author of the BCD PR doesn't know or doesn't care which feature that key belongs to and does not tag it.

> - A few days later, that data is released by BCD. That triggers the next `@mdn/browser-compat-data` Dependabot upgrade PR. After that PR merges, we run the "Update draft features" workflow, which makes the new keys to appear in the `features/draft` folder.

Reviewer 3: ```suggestion
- A few days later, that data is released by BCD. That triggers the next `@mdn/browser-compat-data` Dependabot upgrade PR in the web-features repository. After that PR merges, we run the "Update draft features" workflow, which makes the new keys to appear in the `features/draft` folder.
```

- At this point, a web-features maintainer can assign the key to an existing feature or author a new feature.
- A few days later, that data is released by web-features. When BCD upgrades `web-features`, it'll automatically apply tags to each feature according to the `compat_features` list.

## Circular changes

To prevent circular changes:

- BCD contributors must never create a new tag before the corresponding ID appears in web-features.

> - web-features contributors must never manually remove a `compat_features` list from a feature.

Reviewer: Are both of these enforced by lint/tests? It might be worth mentioning that, if they are.

> - A few days later, that data is released by web-features. When BCD upgrades `web-features`, it'll automatically apply tags to each feature according to the `compat_features` list.

Reviewer 3: Tags get applied to *keys* in BCD, right?

Also, does "apply tags" mean: "add" or "add and remove" (i.e. "synchronize")?

```suggestion
- A few days later, that data is released by web-features. When BCD upgrades `web-features`, it'll automatically adds/removes tags to/from each key according to the `compat_features` list.
```

> - a new key appears in BCD (e.g., via a [Collector](https://github.com/openwebdocs/mdn-bcd-collector PR).
> - The author of the BCD PR doesn't know or doesn't care which feature that key belongs to and does not tag it.
> - A few days later, that data is released by BCD. That triggers the next `@mdn/browser-compat-data` Dependabot upgrade PR in the web-features repository. After that PR merges, we run the "Update draft features" workflow, which makes the new keys to appear in the `features/draft` folder.
> - At this point, a web-features maintainer can assign the key to an existing feature or author a new feature.
> - A few days later, that data is released by web-features. When BCD upgrades `web-features`, it'll automatically add/remove tags to/from each key according to the `compat_features` list.

Reviewer 2: Again, making the linearity more apparent here:

```suggestion
1. A new key appears in BCD (e.g., via a [Collector](https://github.com/openwebdocs/mdn-bcd-collector PR).
2. The author of the BCD PR doesn't know or doesn't care which feature that key belongs to and does not tag it.
3. A few days later, that data is released by BCD. That triggers the next `@mdn/browser-compat-data` Dependabot upgrade PR in the web-features repository.
4. After the Dependabot PR merges, we run the "Update draft features" workflow, which makes the new keys appear in the `features/draft` folder.
5. At this point, a web-features maintainer can assign the key to an existing feature or author a new feature.
6. A few days later, that data is released by web-features. When BCD upgrades `web-features`, it'll automatically add/remove tags to/from each key according to the `compat_features` list.
```

> - BCD contributors must never create a new tag before the corresponding ID appears in web-features. (linters in BCD will prevent this from happening)

Reviewer 3: Worth linking to the documentation that explains how to suggest a new feature?

> To ensure ongoing completeness, web-features maintainers ought to strive to keep generated drafts to an absolute minimum.

Reviewer 2: Conciseness:

```suggestion
To ensure ongoing completeness, web-features maintainers must keep generated drafts to an absolute minimum.
```
